### ISSUE TITLE
[RFC] SentinelOne pulling Scripts data 

### ISSUE THREAD

**Issue by tomsonpl** [elastic/integrations#9313]
As a Kibana user I would like to have more data coming from S1/S1 CloudFunnel integration. I'll try to explain what we are trying to achieve, would be awesome if you could take a quick look and see if this makes sense and if it is possible to be done (without too much trouble). 

1. SentinelOne EDR actions enable users to `execute` scripts - `POST /web/api/v2.1/remote-scripts/execute` 
2. Then these script actions status (we consider it response) by calling `GET /web/api/v2.1/remote-scripts/status` 
3. Then in order to see the script's result we have to call - `POST /web/api/v2.1/remote-scripts/fetch-files` which returns a zip file. 
4. We would like to read the file (*.txt) and save it's content as `result` in ES. 

We are thinking whether it makes sense to pull this data at the integration (ingestion?) level and then in Kibana just operate on our ES data, rather than having to call SentinelOne API from Kibana and save to ES then... 

If you could tell us if any of these 1,2,4 (3 response contains a link that expires so we can't make any use of it saved) points makes sense and if it's doable, that would be great starting point for more discussion's that we have in our team considering SentinelOne.

Thank you!

**Comment by tomsonpl** [comment]
@elastic/security-service-integrations hey team 👋 
We're getting more and more advanced with SentinelOne bi-directional actions. Would be very helpful if you could help us understand what from the scope of this ticket is possible to be done, so we could investigate a way forward :) 
Thank you! 

cc: @jamiehynds

**Comment by chrisberkhout** [comment]
Hi @tomsonpl,

Can you please clarify some things for me...

> 1. SentinelOne EDR actions enable users to execute scripts - POST /web/api/v2.1/remote-scripts/execute

I'm not very familiar with the functionality we have in Elastic Security for user-initiated outgoing actions. Are you asking whether it's possible to let users do this from Kibana, or is the plan to still initiate it from SentinelOne EDR?

> If you could tell us if any of these 1,2,4 (**3 response contains a link that expires so we can't make any use of it saved**) points makes sense

Is the response from (3) `POST /web/api/v2.1/remote-scripts/fetch-files` a `zip` file containing a `txt` file that we want to index? If so, what is the link that expires? I'm wondering if you meant that the status endpoint (2) returns an expiring link, and whether that's used for fetching the files.

---

More generally, at this point I'd say that it could make sense to use an integration to ingest the script results (and perhaps status), if some conditions are met.

Condition 1: It's tolerable to continuously poll for that data. Most of our integrations will make requests at intervals (typical interval times range from 15s to 24h). To poll more frequently or only when needed could be a problem.

Condition 2: The integration can figure out how to fetch the data by itself. If it can call an endpoint to list recently executed scripts and get status and output zips for them, that's good. If you need an ID from the execute response to be able to do the `status` and `fetch-files` requests, that might not work as well.

What do you think about these conditions?

**Comment by tomsonpl** [comment]
Hey @chrisberkhout, thank you for taking a look at this :) 
To give you more context - we have been working on enabling S1 users to call these `actions` directly from Kibana. When they do this, the data about this actions are saved in S1 - eg the execute actions. What we would like to know is if it's possible for integration to be polling these data (1), it's statuses (2), and file content (4) and save to ES. So basically we would like to get the equivalent of data received through API (urls provided above) through integration. 

Regarding the (3) - I think I confused this: 
The request fetches (generates?) file urls, that we later can use to actually download the file. And these urls expire after 5 mins. So it feels like - if integration could fetch file urls, then fetch file and save it's content as txt? On second thought it sounds like an overhead to me, but maybe you have some tricks that would make it feasible? :) 

Condition 1 feels totally fine for us. 
Condition 2 sounds good too, but I am afraid we would need the ID's for some requests. 

cc: @paul-tavares You've been investigating the issue further, is what I explained still valid - or anything changed? Thanks!

**Comment by chrisberkhout** [comment]
Okay, that sounds promising.

It would be good to confirm a way of doing those requests. Chaining together a series of requests wouldn't be a problem. For example, it could lists scripts, then get execution status(es) for each script, then get the output file URLs for each execution, then get each actual file.

I had a quick look for API documentation. It looks like it might be at https://endpoint.sentinelone.net/api-doc/overview but I got redirected to log in for that.

**Comment by tomsonpl** [comment]
Hey, this is the address: https://usea1-partners.sentinelone.net/api-doc/overview I created an account for you, check your email :)

**Comment by chrisberkhout** [comment]
Thanks. I had a look at the documentation. It looks like probably an ID returned by the Run Remote Script call is needed for the Get Script Results call. It's hard to tell because there are quite a few types of ID: `pendingExecutionId`, `parentTaskId`, `taskId`, (script) `id`, `scheduledTaskId`.

It looks like the "guardrail configuration" lets you require approval of pending executions and pending executions can be listed. That returns a `scheduledTaskId`, but that might not be enough for the Get Remote Scripts Tasks Status call or the Get Script Results call.

The next step would be to write a series of curls that runs through this process and tries to make it work without having to pass data from Run Remote Script call to the later Get Script Results call.

If getting an ID from the Run Remote Script call is the only way, an integration could perhaps query Elasticsearch for current IDs before talking to SentinelOne. I'm not sure whether we've used that pattern in another integration.

**Comment by paul-tavares** [comment]
Hey @chrisberkhout ,
Sorry for the delay.

Before I detail the needs around pulling in SentinelOne script execution data, I would like to see if the following is possible for the existing data that is being pulled from SentinelOne:

- For every record streamed to ES by the integration, if the data is associated with a SentinelOne agent, can we include the agent's `id`?
    - The `sentinel_one.[source_type].agent.id` already seems to be part of some of the data streamed to ES - example:
        - for the activity index, its captured under `sentinel_one.activity.agent.id`
        - the Threats index its captured under `sentinel_one.threat.agent.id`
        - the `alerts` data for S1 does not include a `sentinel_one.alert.agent.id`
            - it includes only a `sentinel_one.alert.agent.site_id`
            - Instead, this index data has a document top-level property named  `observer.serial_number` that seems to be the agent's `uuid`. We been using this to identify the agent associated with the alert.
        - We would like to move to using `agent.id` and thus having that on the alert data (and other data if currently missing) would be very helpful
        - FYI: the `agent.uuid` in S1 can be changed via the API (and UI), however, the `agent.id` seems to be non-mutable, thus why we would like to move to using this instead in the functionality we are working on
- Can we change the default intervals for the data pulling to `30s` for all data? We are building functionality for our users that looks at the data the integration streamed to ES and it would be ideal to minimize the time the user has to wait for a response. 


Let us know if you rather us create a new issue to track this request ☝️ 


</br>
</br>

______

Ok. On to the original request in this issue. 

> [!NOTE]
> **TL;DR:** now that I completed writing the below content: I don't think its going to be possible to pull in script/task status 😞 ... but read on:



## Retrieving script run information from SentinelOne. 

As @tomsonpl indicated, in security we are currently providing our users with the ability to respond to alerts created in our SIEM from SentinelOne data. Things like `isolate` the host, `kill-process`, `get-file`, etc. We send these action to S1 via a Stack Connector and then check the data gathered and streamed to ES by the integration policy to report on status of those. Some actions in SentinelOne are achieved by running SentinelOne scripts, which SentinelOne seems to schedule via a Task. 

At this time, I don't think we need the files associated with script executions to actually be stored in ElasticSearch - we have found a way to support file download via Kibana since this issue was created by Tomasz. What we would like to see perhaps captures and streamed to ES is the Script/Task execution status data


### Determine list of scripts/task created (since last time data was pulled)

I'm not seeing a good way yet to do this. There is an API used by the SentinelOne UI that seems to return the data we need, but that API seems to be `private` given the route path: 

```
https://usea1-partners.sentinelone.net/web/api/v2.1/private/bulk-tasks?limit=100&sortBy=createdAt&sortOrder=desc&createdAt__gt=1715098954144&accountIds=1392053568574369781
```

_(Note the `/private/` in the route 😞 )_


<details><summary>Response from the private `bulk-tasks` api</summary>

```jsonp
{
    "data": [
        {
            "canceledTaskInScopeCounter": 0,
            "completedTaskInScopeCounter": 1,
            "createdAt": "2024-05-14T15:54:48.900387Z",
            "description": "ptavares-process - test",
            "expiredTaskInScopeCounter": 0,
            "failedTaskInScopeCounter": 0,
            "id": "1950127291056776620",      // <<<<<< This is the `parentTaskId` we would need
            "inProgressTaskInScopeCounter": 0,
            "initiatedBy": "Paulo Tavares",
            "initiatedById": "1809444483386312727",
            "partiallyCompletedTaskInScopeCounter": 0,
            "pendingTaskInScopeCounter": 0,
            "pendingUserActionTaskInScopeCounter": 0,
            "scheduledTaskInScopeCounter": 0,
            "tasksInScope": 1,
            "totalTasks": 1,
            "type": "script_execution",
            "updatedAt": "2024-05-14T15:55:12.032835Z"
        }
    ],
    "pagination": {
        "nextCursor": null,
        "totalItems": 1
    }
}
```

</details>


#### Maybe get it from the SentinelOne activity log?

The other place I thought of was the Activity log. The API call looks like this and it would identify the newly initiated scripts:

```
https://usea1-partners.sentinelone.net/web/api/v2.1/activities?activityTypes=3618&sortBy=createdAt&sortOrder=desc

```

The key filter above is the `activityTypes=3618` - this activity type is used when a script action is initiated:

```
{
      "action": "Script Action Initiated",
      "descriptionTemplate": "The management user {{username}} executed the {{script_name}} script which will apply to a total of {{number}} endpoints. {% if password_from_scope_used %} Execution artifacts are protected by Remote Ops password from {{password_scope_level}} scope {{ password_scope_name }}.{% endif %}",
      "id": 3618
    },
```

The response, however, does not include the data we would need (`parentTaskId`) to then call the script status API. I think its because the activity log entry is done prior to the task actually being created 🤷 .


<details><summary>Example of Activity log entry</summary>

```jsonp
{
			"accountId": "1392053568574369781",
			"accountName": "Elastic",
			"activityType": 3618,
			"activityUuid": "8841911a-78a2-418e-b31d-f334906df446",
			"agentId": null,
			"agentUpdatedVersion": null,
			"comments": null,
			"createdAt": "2024-05-14T15:54:48.919093Z",
			"data": {
				"accountName": "Elastic",
				"creator": "paulo.tavares@elastic.co",
				"externalServiceId": null,
				"fileName": "19488/None/1392053568574369781/qnfn5c2TbknjzY2YgDBKe%2F3e5Ts8hsKauLdaLdCDItw%3D/ptavares-processes.sh",
				"fullScopeDetails": "Account Elastic",
				"fullScopeDetailsPath": "Global / Elastic",
				"groupName": null,
				"ipAddress": "108.77.84.191",
				"isAvailableForArs": false,
				"isInputRequired": false,
				"mgmtId": 19488,
				"number": 1,
				"osTypes": [
					"linux"
				],
				"outputDestination": "None",
				"passwordFromScopeUsed": false,
				"realUser": null,
				"requiresApproval": false,
				"scopeId": 1392053568574369800,
				"scopeLevel": "Account",
				"scopeName": "Elastic",
				"scriptFileName": "ptavares-processes.sh",
				"scriptFileSize": 186,
				"scriptId": "1891514995700968232",
				"scriptName": "ptavares-process",
				"scriptRuntimeTimeoutSeconds": 3600,
				"scriptType": "dataCollection",
				"signature": "qnfn5c2TbknjzY2YgDBKe/3e5Ts8hsKauLdaLdCDItw=",
				"signatureType": "SHA-256",
				"siteName": null,
				"sourceType": "UI",
				"taskDescription": "ptavares-process - test",
				"type": "dataCollection",
				"userId": 1809444483386312700,
				"username": "Paulo Tavares",
				"version": "1.0.0"
			},
			"description": null,
			"groupId": null,
			"groupName": null,
			"hash": null,
			"id": "1950127291232937389",
			"osFamily": null,
			"primaryDescription": "The management user Paulo Tavares executed the ptavares-process script which will apply to a total of 1 endpoints. ",
			"secondaryDescription": "IP address: 108.77.84.191",
			"siteId": null,
			"siteName": null,
			"threatId": null,
			"updatedAt": "2024-05-14T15:54:48.916571Z",
			"userId": "1809444483386312727"
		}
``` 


</details>



___________

So... after all that, I don't think we will be able to actually retrieve this data via the integration into an ES index.

What we will end up doing (which is ok) is add code on our side (Kibana) to actually call the S1 API to get the status of the script _(we will be storing the `ParentTaskId` at the time we send the request to SentinelOne's API)_

**Comment by chrisberkhout** [comment]
Hi @paul-tavares,

Thanks for that. Just confirming my understanding...

I don't know the details on the Kibana side, but in general given this is about managing a task initiated by the user from in Kibana, it makes sense to me that Kibana does the status tracking. It sounds like that is a better option than the alternatives:
- Use the `.../private/bulk-tasks` endpoint to list `parentTaskId`s.
- Have Kibana store the `parentTaskId` in Elasticsearch and have the integration poll Elasticsearch for the IDs and poll the S1 API for status updates.

For the agent ID mapping and interval changes it would be good to open a separate issue. I think there's some field details to confirm but it shouldn't be a problem to do.

**Comment by paul-tavares** [comment]
Your understanding is correct @chrisberkhout . Thanks for the time on this. I will open a new issue for the `agent.id` and interval change request.

**Comment by paul-tavares** [comment]
FYI: New issue opened: https://github.com/elastic/integrations/issues/9879
